examples/GraphTraversal/Graph.roc

```diff
@@ -0,0 +1,213 @@
+## The Graph interface represents a graph using an adjacency list and exposes
+## functions for working with graphs, such as creating one from a list and
+## performing a depth-first or breadth-first search.
+interface Graph
+    exposes [
+        Graph,
+        fromList,
+        fromDict,
+        dfs,
+        bfs,
+    ]
+    imports []
+
+## Graph type representing a graph as a dictionary of adjacency lists
+## where each key is a vertex and each value is a list of its adjacent vertices.
+Graph a := Dict a (List a) | a has Eq
+
+## Create a Graph from an adjacency list.
+fromList : List (a, List a) -> Graph a
+fromList = \adjacencyList ->
+    emptyDict = Dict.withCapacity (List.len adjacencyList)
+
+    update = \dict, (vertex, edges) ->
+        Dict.insert dict vertex edges
+
+    List.walk adjacencyList emptyDict update
+    |> @Graph
+
+## Create a Graph from an adjacency list.
+fromDict : Dict a (List a) -> Graph a
+fromDict = @Graph
+
+## Perform a depth-first search on a graph to find a target vertex.
+##
+## - `isTarget` : A function that returns true if a vertex is the target.
+## - `root`     : The starting vertex for the search.
+## - `graph`    : The graph to perform the search on.
+dfs : (a -> Bool), a, Graph a -> Result a [NotFound]
+dfs = \isTarget, root, @Graph graph ->
+    dfsHelper isTarget [root] (Set.empty {}) graph
+
+# A helper function for performing the depth-first search.
+#
+# `isTarget` : A function that returns true if a vertex is the target.
+# `stack`    : A List of vertices to visit.
+# `visited`  : A Sist of visited vertices.
+# `graph`    : The graph to perform the search on.
+dfsHelper : (a -> Bool), List a, Set a, Dict a (List a) -> Result a [NotFound]
+dfsHelper = \isTarget, stack, visited, graph ->
+    when stack is
+        [] ->
+            Err NotFound
+
+        [.., current] ->
+            rest = List.dropLast stack
+
+            if isTarget current then
+                Ok current
+            else if Set.contains visited current then
+                dfsHelper isTarget rest visited graph
+            else
+                newVisited = Set.insert visited current
+
+                when Dict.get graph current is
+                    Ok neighbors ->
+                        # filter out all seen neighbors
+                        filtered =
+                            neighbors
+                            |> List.keepIf (\n -> !(Set.contains visited n))
+                            |> List.reverse
+
+                        # newly explored nodes are added to LIFO stack
+                        newStack = List.concat rest filtered
+
+                        dfsHelper isTarget newStack newVisited graph
+
+                    Err KeyNotFound ->
+                        dfsHelper isTarget rest newVisited graph
```

Comment:
My comment earlier on filtering the neighbors and adding the stack elements was a bit more than what you have here. Something like:
```suggestion
dfsHelper = \isTarget, stack, seen, graph ->
    when stack is
        [] ->
            Err NotFound

        [.., current] ->
            rest = List.dropLast stack

            if isTarget current then
                Ok current
            else
                when Dict.get graph current is
                    Ok neighbors ->
                        # filter out all seen neighbors
                        filtered =
                            neighbors
                            |> List.keepIf (\n -> !(Set.contains seen n))
                            |> List.reverse

                        newSeen = Set.union seen (Set.fromList filtered)

                        # newly explored nodes are added to LIFO stack
                        newStack = List.concat rest filtered

                        dfsHelper isTarget newStack newSeen graph

                    Err KeyNotFound ->
                        dfsHelper isTarget rest seen graph
```

Note: `Set.union seen (Set.fromList filtered)` may not be the most efficient. It may be better to to insert each element from the list and not build the second set.

Note: you also need to update the starting `seen` list to be `Set.single root` instead of empty.

Some would apply to bfs.